### PR TITLE
Adds command xrustc

### PR DESCRIPTION
**Pull request by qqwa**
Some options can only be passed to rustc with arguments e.g. `--emit=link=path/exe`. I'm not sure if it will break anything for others who would use `cargo build` instead of `cargo rustc`

### CODE REVIEW

**Comment by phil-opp** [comment]
Thanks for the PR! Support for `cargo rustc` building is something that we should definitely do.

> I'm not sure if it will break anything for others who would use cargo build instead of cargo rustc

I don't know either. So how about adding a separate `cargo xrustc` executable that invokes `cargo rustc`? This would have the big advantage that we just need to remove an `x` when `cargo` becomes able to build the sysroot itself.

**Comment by qqwa** [comment]
Fine by me. I pushed a commit which splits it in two binaries. We probably want to move more stuff from the binaries to the lib. And minimize the imports, currently it's for all the same, just that the binaries also include xargo_lib.

File: src/bin/cargo-xrustc.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+use xargo_lib::xargo;
+
+use xargo_lib::CurrentDirectory;
+use xargo_lib::CompilationMode;
+
+const HELP: &str = include_str!("help.txt");
+
+
+
+pub fn main() {
```
Comment (phil-opp) [inline src/bin/cargo-xrustc.rs]: I think most of the `main`, `run`, and `build` methods could be moved to the lib (same for the `cargo-build` binary`).

**Comment by qqwa** [comment]
I left the main as it was, because its only error handling, which shouldn't change that much anyways.

And the help text for `cargo xrustc` will use the same as `cargo xbuild`. Should i add new one for xrustc or can we use something with `!format`?

**Comment by qqwa** [comment]
This would work to show the correct command name.
```
pub fn main() {
    fn show_backtrace() -> bool {
        env::var("RUST_BACKTRACE").as_ref().map(|s| &s[..]) == Ok("1")
    }

match xargo_lib::run("xrustc", &format!(include_str!("help.txt"), "xrustc"), true) {
```

Or inside the run function, something like this:
```
pub fn run(command_name: &str, help: &str, use_rustc: bool) -> Result<Option<ExitStatus>> {
    use cli::Command;

    let (command, args) = cli::args(command_name)?;
    match command {
        Command::Build => Ok(Some(build(args, use_rustc)?)),
        Command::Help => {
            print!("{}", help.replace("{command_name}", command_name);
            Ok(None)
        }
```

**Comment by phil-opp** [comment]
Thanks for the update!

I would like to move the `main` function to the lib too, e.g. as `main_common(command_name: &str)`, because maybe we want to change the error handling. And it's an easy way to avoid the code duplication.

It would be cool to use a common `help.txt`, but note that we would also need to replace the `cargo build` instances in the text.

**Comment by qqwa** [comment]
I also removed the `use_rustc` bool and just pass the command name to cargo.

The help message should now also be correct for any command.

File: src/cli.rs [hunk 2]
```diff
@@ -27,10 +27,10 @@ impl Args {
     }
 }
 
-pub fn args() -> Result<(Command, Args), String> {
+pub fn args(command_name: &str) -> Result<(Command, Args), String> {
     let mut args = env::args().skip(1);
-    if args.next() != Some("xbuild".into()) {
-        Err("must be invoked as cargo subcommand: `cargo xbuild`")?;
+    if args.next() != Some("x".to_string() + command_name) {
+        Err(format!("must be invoked as cargo subcommand: `cargo {}`", command_name))?;
```
Comment (phil-opp) [inline src/cli.rs]: Shouldn't it be `cargo x{}` (with prepended `x`)?

File: src/help.txt [hunk 3]
```diff
@@ -1,14 +1,14 @@
-Wrapper for `cargo build` that cross-compiles core, compiler_builtins and alloc
+Wrapper for `cargo {command_name}` that cross-compiles core, compiler_builtins and alloc
```
Comment (phil-opp) [inline src/help.txt]: Nice solution using named formatting parameters!

**Comment by phil-opp** [comment]
@qqwa Could you also update the title of this PR?

File: src/cli.rs (hunk 2 above)
Comment (qqwa) [reply src/cli.rs]: Yeah I missed that one.

**Comment by qqwa** [comment]
You can squash it^^

**Comment by phil-opp** [comment]
Thanks a lot!

**Comment by phil-opp** [comment]
Released as version 0.4.8

**Pull request merged**